packages/zoe/src/contracts/newSwap/getCurrentPrice.js

```diff
@@ -0,0 +1,260 @@
+// @ts-check
+
+import '../../../exported';
+
+import { assert, details as X } from '@agoric/assert';
+import { amountMath } from '@agoric/ertp';
+import { multiplyBy, makeRatio, natSafeMath } from '../../contractSupport';
+
+const { ceilDivide } = natSafeMath;
+const BASIS_POINTS = 10000;
+
+/**
+ * Build functions to calculate prices for multipoolAutoswap. Four methods are
+ * returned, as two complementary pairs. In one pair of methods the caller
+ * specifies the amount they will pay, and in the other they specify the amount
+ * they wish to receive. The two with shorter names (getOutputForGivenInput,
+ * getInputForGivenOutput) are consistent with the uniswap interface. They each
+ * return a single amount. The other two return { amountIn, centralAmount,
+ * amountOut }, which specifies the best exchange consistent with the request.
+ * centralAmount is omitted from these methods' results in the publicFacet.
+ *
+ * @param {(brand: Brand) => boolean} isSecondary
+ * @param {(brand: Brand) => boolean} isCentral
+ * @param {(brand: Brand) => Pool} getPool
+ * @param {Brand} centralBrand
+ * @param {bigint} poolFeeBP - fee (BP) to the pool in every swap.
+ * @param {bigint} protocolFeeBP - fee (BP) charged in RUN to every swap.
+ */
+
+export const makeGetCurrentPrice = (
+  isSecondary,
+  isCentral,
+  getPool,
+  centralBrand,
+  poolFeeBP,
+  protocolFeeBP,
+) => {
+  const protocolFeeRatio = makeRatio(protocolFeeBP, centralBrand, BASIS_POINTS);
+  const halfPoolFeeBP = ceilDivide(poolFeeBP, 2);
+
+  const getPriceGivenAvailableInput = (amountIn, brandOut) => {
+    const { brand: brandIn } = amountIn;
+
+    if (isCentral(brandIn) && isSecondary(brandOut)) {
+      // we'll charge the protocol fee before sending the remainder to the pool,
+      // so we need to subtract it from amountIn before getting a quote, and add
+      // it back before sending out the quote.
+      const prelimProtocolFee = multiplyBy(amountIn, protocolFeeRatio);
+      const poolAmountIn = amountMath.subtract(amountIn, prelimProtocolFee);
+      const price = getPool(brandOut).getPriceGivenAvailableInput(
+        poolAmountIn,
+        brandOut,
+        poolFeeBP,
+      );
+      const protocolFee = multiplyBy(price.amountIn, protocolFeeRatio);
+      const amountInFinal = amountMath.add(price.amountIn, protocolFee);
+
+      return {
+        amountIn: amountInFinal,
+        amountOut: price.amountOut,
+        protocolFee,
+      };
+    } else if (isSecondary(brandIn) && isCentral(brandOut)) {
+      // We'll charge the protocol fee after getting the quote
+      const price = getPool(brandIn).getPriceGivenAvailableInput(
+        amountIn,
+        brandOut,
+        poolFeeBP,
+      );
+      const protocolFee = multiplyBy(price.amountOut, protocolFeeRatio);
```

Comment:
I'm not sure I like the interaction between the pool fee the protocol fee here. We're charging the protocol fee on an amountOut that has already been reduced by the pool fee. So it will not actually be the protocolFeeRatio we are charging. Can we rethink this to ensure that we know exactly what percentage we are charging?

Reply:
My goal here is not to match uniswap behavior, but to match our own goals: if we are charging x%, I want to make sure we're charging the amount we say. Right now, we're not charging the amounts we say, as far as I can tell, because we are charging based on an amount that has already had a fee taken out. It seems much easier to drop the old fee calculation entirely, charge fees once per operation, and then divy up the fees according to their use. 

> We had previously concluded that the way to charge protocol fees consistently in RUN was to subtract the RUN amount before using the uniswap formulas to convert deltaX to deltaY or vice versa. 

Hmm, but you are still using the old fee calculation functions, which we didn't agree on. I don't think we should be using the same old bondingCurve.js functions with a protocol fee added on top. That doesn't make sense to me, given the added complexity.